src/realm/sync/noinst/client_history_impl.cpp

```diff
@@ -1302,119 +1156,27 @@ void ClientHistoryImpl::record_current_schema_version(Array& schema_versions, ve
-        Array cooked_history{m_shared_group->get_alloc()};
-        cooked_history.set_parent(&root, s_cooked_history_iip);
-        // We should have no cooked history in existing Realms.
-        REALM_ASSERT(cooked_history.get_ref_from_parent() == 0);
+        m_arrays->init_from_ref(ref);
     }
     else {
-        REALM_ASSERT(!m_ct_history);
-        REALM_ASSERT(!m_changesets);
-        REALM_ASSERT(!m_reciprocal_transforms);
-        REALM_ASSERT(!m_remote_versions);
-        REALM_ASSERT(!m_origin_file_idents);
-        REALM_ASSERT(!m_origin_timestamps);
-        Allocator& alloc = m_shared_group->get_alloc();
-        std::unique_ptr<Arrays> arrays(new Arrays(alloc)); // Throws
-        arrays->root.init_from_ref(ref);
-        if (m_group)
-            gf::set_history_parent(*m_group, arrays->root);
-        std::unique_ptr<BinaryColumn> ct_history;
-        {
-            ct_history.reset(new BinaryColumn(alloc)); // Throws
-            ct_history->set_parent(&arrays->root, s_ct_history_iip);
-            ct_history->init_from_parent();
-        }
-        std::unique_ptr<BinaryColumn> changesets;
-        {
-            changesets.reset(new BinaryColumn(alloc)); // Throws
-            changesets->set_parent(&arrays->root, s_changesets_iip);
-            changesets->init_from_parent(); // Throws
-        }
-        std::unique_ptr<BinaryColumn> reciprocal_transforms;
-        {
-            reciprocal_transforms.reset(new BinaryColumn(alloc)); // Throws
-            reciprocal_transforms->set_parent(&arrays->root, s_reciprocal_transforms_iip);
-            reciprocal_transforms->init_from_parent();
-        }
-        std::unique_ptr<IntegerBpTree> remote_versions;
-        {
-            remote_versions.reset(new IntegerBpTree(alloc)); // Throws
-            remote_versions->set_parent(&arrays->root, s_remote_versions_iip);
-            remote_versions->init_from_parent();
-        }
-        std::unique_ptr<IntegerBpTree> origin_file_idents;
-        {
-            origin_file_idents.reset(new IntegerBpTree(alloc)); // Throws
-            origin_file_idents->set_parent(&arrays->root, s_origin_file_idents_iip);
-            origin_file_idents->init_from_parent();
-        }
-        std::unique_ptr<IntegerBpTree> origin_timestamps;
-        {
-            origin_timestamps.reset(new IntegerBpTree(alloc)); // Throws
-            origin_timestamps->set_parent(&arrays->root, s_origin_timestamps_iip);
-            origin_timestamps->init_from_parent();
-        }
-        m_arrays = std::move(arrays);
-        m_ct_history = std::move(ct_history);
-        m_changesets = std::move(changesets);
-        m_reciprocal_transforms = std::move(reciprocal_transforms);
-        m_remote_versions = std::move(remote_versions);
-        m_origin_file_idents = std::move(origin_file_idents);
-        m_origin_timestamps = std::move(origin_timestamps);
+        m_arrays.emplace(m_db->get_alloc(), *m_group, ref);
```

Comment:
@tgoyne This change assumes there is always a group, but that was not the case before. We have a crash because of this exact reason (https://github.com/realm/realm-core/issues/7041), and I'm trying to figure out if this change introduced a bug, or if it existed before (and we perhaps don't call `set_group()` accordingly)

Reply:
`set_group()` is called by `Replication::initiate_transact()` as part of beginning a write transaction, and `prepare_for_write()` can only be called inside a write transaction. `m_group` being nullptr here suggests that a function which writes to the Realm is being called without the DB having an active write transactions.

Reply:
`update_from_ref_and_version` is called from `internal_advance_read` after the write lock is acquired but before `initiate_transact` in the linked issue #7041. The code in `Transaction` was doing the same thing back then. Was that also taken into account somehow? From what i can tell this may fail if `internal_advance_read` actually moves to the latest version available.